### DISCUSSION TITLE
Who can toggle an NFT transfer?

### DISCUSSION THREAD

**Discussion by gabrielstoica** [sablier-labs/evm-monorepo#1050] [Q&A] [answered]
Hey, 

I'm looking for a way to withdraw a stream and transfer it to another recipient in a programmatic way, through an intermediary contract (aka Stream Management contract - see [this discussion](https://github.com/sablier-labs/v2-core/discussions/977)).
As stated in [this discussion](https://github.com/sablier-labs/v2-core/discussions/669) on NFT transferability, only the stream recipient can toggle the transfer - through the `withdrawMaxAndTransfer` method which [checks](https://github.com/sablier-labs/v2-core/blob/main/src/abstracts/SablierV2Lockup.sol#L424-L428) if the `msg.sender` is the current recipient.

However, this can be bypassed through a 3-step process, as follows:
1. Approve the intermediary contract to transfer the stream's NFT by calling the `SablierV2Lockup.approve` method;
2. Withdraw the maximum withdrawable amount from the stream by calling the `withdrawMax` method;
3. Transfer the NFT by calling the `transferFrom` exposed by the `SablierV2Lockup` implementation;

With all that said, should only the stream recipient be able to transfer the NFT? If so, the workflow above should not be allowed. Otherwise, if that's not an issue, is there another way to accomplish it?

Thanks!

**Comment by PaulRBerg** [comment, marked as answer]
Hi Gabriel, thanks for your question.

If the user approves an operator through the ERC-721 `approve` method, then it is a perfectly legitimate behavior for the operator to be able to (i) withdraw the maximum withdrawable amount from the stream and (ii) transfer the NFT.

This makes me think that maybe we should rewrite the check [here](https://github.com/sablier-labs/v2-core/blob/f84a53de1b6d51b62ac205ac38c6a9721419ba2e/src/abstracts/SablierV2Lockup.sol#L425-L428) to call `_isCallerStreamRecipientOrApproved` instead of checking just the recipient.

cc @sablier-labs/solidity

> is there another way to accomplish it?

No, at the moment, you would have to call the two functions separately. But, as I've said, it should be fine to modify the code to allow this behavior to be enacted through the `withdrawMaxAndTransfer` function directly. I have created a GitHub issue for this.

Thanks for your feedback!

**Comment by gabrielstoica** [comment]
Hi Paul,

Thanks for your prompt answer. That make sense!

> This makes me think that maybe we should rewrite the check [here](https://github.com/sablier-labs/v2-core/blob/f84a53de1b6d51b62ac205ac38c6a9721419ba2e/src/abstracts/SablierV2Lockup.sol#L425-L428) to call _isCallerStreamRecipientOrApproved instead of checking just the recipient.

That's what I was going to ask & propose next. 

Thanks again!

**Reply by PaulRBerg** [reply]
I thank you!
